packages/grid/_modules_/grid/hooks/features/column-reorder/useColumnReorder.tsx

```diff
@@ -50,45 +50,59 @@ const hasCursorPositionChanged = (
 ): boolean =>
   currentCoordinates.x !== nextCoordinates.x || currentCoordinates.y !== nextCoordinates.y;
 
-export const useColumnReorder = (columnsRef: React.RefObject<HTMLDivElement>, apiRef: ApiRef) => {
+export const useColumnReorder = (apiRef: ApiRef): void => {
   const logger = useLogger('useColumnReorder');
 
-  const dragCol = React.useRef<ColDef | null>(null);
+  const [, setGridState] = useGridState(apiRef);
+  const dragCol = useGridSelector(apiRef, columnReorderDragColSelector);
   const dragColNode = React.useRef<HTMLElement | null>(null);
+  const columnsHeaderRef = React.useRef<HTMLElement | null>(null);
   const cursorPosition = React.useRef<CursorCoordinates>({
     x: 0,
     y: 0,
   });
   const removeDnDStylesTimeout = React.useRef<number>();
 
   const handleDragEnd = React.useCallback((): void => {
-    logger.debug(`End dragging col ${dragCol.current!.field}`);
+    logger.debug(`End dragging col ${dragCol!.field}`);
     apiRef.current.publishEvent(COL_REORDER_STOP);
 
     clearTimeout(removeDnDStylesTimeout.current);
 
-    columnsRef.current!.classList.remove(HEADER_CELL_DROP_ZONE_CSS_CLASS);
+    columnsHeaderRef.current!.classList.remove(HEADER_CELL_DROP_ZONE_CSS_CLASS);
     dragColNode.current!.parentElement!.classList.remove('MuiDataGrid-colCellMoving');
     dragColNode.current!.removeEventListener(DRAGEND, handleDragEnd);
-    dragCol.current = null;
     dragColNode.current = null;
-  }, [columnsRef, apiRef, logger]);
+
+    setGridState((oldState) => {
+      return {
+        ...oldState,
+        columnReorder: { ...oldState.columnReorder, dragCol: null },
+      };
+    });
+  }, [apiRef, dragCol, setGridState, logger]);
 
   const handleDragStart = React.useCallback(
     (col: ColDef, currentTarget: HTMLElement): void => {
       logger.debug(`Start dragging col ${col.field}`);
       apiRef.current.publishEvent(COL_REORDER_START);
 
-      dragCol.current = col;
       dragColNode.current = currentTarget;
-      dragColNode.current.addEventListener(DRAGEND, handleDragEnd, { once: true });
       dragColNode.current.classList.add(HEADER_CELL_DRAGGING_CSS_CLASS);
       dragColNode.current.parentElement!.classList.add('MuiDataGrid-colCellMoving');
```

Comment:
Does it mean we can solve #452 after the refactorization? 

```suggestion
```

If we can this sound like this refactorization is bringing value. If we can't, then we need to rethink the state.

Reply:
Done, I've fixed it ;)